kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/LogSegment.java

```diff
@@ -277,4 +343,78 @@ private String formatList(List<FileStatus> list) {
     return "\n    "
         + list.stream().map(FileStatus::toString).collect(Collectors.joining(",\n    "));
   }
+
+  // Class to resolve the final list of deltas + log compactions to return
+  private class LogCompactionResolver {
+    int compactionPos = 0;
+    long currentCompactionHi = -1;
+    long currentCompactionLo = -1;
+    final List<FileStatus> compactionsReversed;
+
+    final Iterator<FileStatus> deltaIt;
+
+    LogCompactionResolver(List<FileStatus> allFilesReversed, List<FileStatus> compactionsReversed) {
+      this.deltaIt = allFilesReversed.iterator();
+      this.compactionsReversed = compactionsReversed;
+    }
+
+    // We have two lists, one of deltas and one of compactions. Each is sorted in DESCENDING
+    // order. Given this, resolves as follows:
+    // - set a "hi/lo" goalpost around the next compactions
+    // - for each delta, if its version is:
+    //   - greater than the current compaction high point, include it, move to next delta
+    //   - less than (but not equal to) the current low point, skip it, move to next delta
+    //   - equal to the current compaction low point, we're above to transition out of the
+    //     compaction, so, include the compaction, find the next compaction that has a high
+    //     point lower than our current low point and set that to the current compaction to
+    //     consider. This deals with overlapping compactions in a greedy way, ensuring we
+    //     ignore any overlapping compactions.
+    List<FileStatus> resolveFiles() {
+      ArrayList<FileStatus> ret = new ArrayList<FileStatus>();
+      setHiLo();
+      while (deltaIt.hasNext()) {
+        FileStatus currentDelta = deltaIt.next();
+        long deltaVersion = FileNames.deltaVersion(currentDelta.getPath());
+        if (deltaVersion == currentCompactionLo) {
+          // we're about to cross out of the compaction. insert the compaction and advance to the
```

Comment:
I think this requires (fairly so) that the delta versions are contiguous. i.e. it requires that 10.json exists for us to filly accept 10.20.json into the `ret`

Q: What would change if you added the compaction at the "start" as opposed to at the end?

Could we future proof this for some crazy future world where there are gaps in the delta log (suppose a future catalogManagedV2 table feature allows for unordered backfill) ?

Not a blocker, but wanted to ask this / call it out

Reply:
Well, if there are gaps I think the logic gets _way_ more complex. If we assume we handle all the missing stuff somehow, I guess the "fix" here would just be `(deltaVersion <= currentCompactionLo)`, so maybe not worth trying to "fix" things until we know what we actually need to do.